go/src/lib.rs

```diff
@@ -1,9 +1,162 @@
+/// * The `conn_ptr` pointer in the returned `ConnectionResponse` must live until it is passed into `close_client`.
+/// * The `connection_error_message` pointer in the returned `ConnectionResponse` must live until the returned `ConnectionResponse` pointer is passed to `free_connection_response`.
+/// * Both the `success_callback` and `failure_callback` function pointers need to live until the client is closed via `close_client` since they are used when issuing Redis commands.
+// TODO: Consider making this async
+#[no_mangle]
+pub unsafe extern "C" fn create_client(
+    connection_request_bytes: *const u8,
+    connection_request_len: usize,
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+) -> *const ConnectionResponse {
+    let request_bytes =
+        unsafe { std::slice::from_raw_parts(connection_request_bytes, connection_request_len) };
+    let response = match create_client_internal(request_bytes, success_callback, failure_callback) {
+        Err(err) => ConnectionResponse {
+            conn_ptr: std::ptr::null(),
+            connection_error_message: CString::into_raw(CString::new(err).unwrap()),
+        },
+        Ok(client) => ConnectionResponse {
+            conn_ptr: Box::into_raw(Box::new(client)) as *const c_void,
+            connection_error_message: std::ptr::null(),
+        },
+    };
+    Box::into_raw(Box::new(response))
+}
+
+/// Closes the given client, deallocating it from the heap.
+///
+/// `client_ptr` is a pointer to the client returned in the `ConnectionResponse` from `create_client`.
+///
+/// # Safety
+///
+/// * `client_ptr` must be obtained from the `ConnectionResponse` returned from `create_client`.
+/// * `client_ptr` must be valid until `close_client` is called.
+/// * `client_ptr` must not be null.
+#[no_mangle]
+pub unsafe extern "C" fn close_client(client_ptr: *const c_void) {
+    let client_ptr = unsafe { Box::from_raw(client_ptr as *mut Client) };
+    let _runtime_handle = client_ptr.runtime.enter();
+}
+
+/// Deallocates a `ConnectionResponse`.
+///
+/// This function also frees the contained error.
+///
+/// # Safety
+///
+/// * `connection_response_ptr` must be obtained from the `ConnectionResponse` returned from `create_client`.
+/// * `connection_response_ptr` must be valid until `free_connection_response` is called.
+/// * `connection_response_ptr` must not be null.
+/// * The contained `connection_error_message` must be obtained from the `ConnectionResponse` returned from `create_client`.
+/// * The contained `connection_error_message` must be valid until `free_connection_response` is called and it must outlive the `ConnectionResponse` that contains it.
 #[no_mangle]
-pub extern "C" fn create_connection() -> *const c_void {
-    todo!()
+pub unsafe extern "C" fn free_connection_response(
+    connection_response_ptr: *mut ConnectionResponse,
+) {
```

Comment:
team suggests we just check if the connection response pointer is null so we don't have to worry about the null condition.

Reply:
Please, update the docs, because pointers could be null